enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,302 @@
+  - "@wking, upgrades team"
+  - "@LalatenduMohanty, upgrades team"
+  - "@soltysh, workloads team"
+approvers:
+  - "@sdodson"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2022-07-29
+last-updated: 2022-07-29
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/USHIFT-233
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1850656
+---
+
+# Advertise Version Details
+
+## Summary
+
+We have several use cases for allowing authenticated users to discover
+whether they are communicating with OCP or MicroShift, including the
+version of the product. We already have the ClusterVersion resource
+for OCP, but that requires special privileges to read and is not
+present in a MicroShift deployment. This enhancement describes a
+standard that both platforms can use to share platform and version
+information with API callers.
+
+## Motivation
+
+Users configuring applications to deploy on different versions of
+OpenShift may need to adjust their configuration to work differently
+on OCP and MicroShift. Ideally this is not needed, but in practice it
+is useful because of different APIs being available and different
+constraints being in place. It is possible for a cluster admin to read
+the ClusterVersion resource in an OCP cluster to discover that it is
+OCP and what version is running. Limiting ourselves to this approach has two
+drawbacks. First, not all console users have cluster admin privileges,
+which means that the console cannot show them the cluster version
+details. Second, the ClusterVersion API is not present in a MicroShift
+deployment, so instead [we created a
+ConfigMap](https://github.com/openshift/microshift/pull/776) in the
+`kube-public` namespace containing some basic version
+details. That means a caller must know two different APIs to check the
+version and product they are communicating with. This enhancement
+describes how we can standardize on one to make it easier for all
+authenticated users to obtain basic version details for a cluster.
+
+### User Stories
+
+As a console user, I want to see the version of OCP I am using so that
+I can refer to the correct version of the documentation and API
+references.
+
+As a console developer, I want to be able to obtain the version of OCP
+that I am using so that the console app can use different
+implementation choices for behaviors that cannot be discovered by
+probing the Kubernetes API versions supported.
+
+As an authenticated API caller, I want to obtain the product name and
+version of the OpenShift variant I am communicating with so that I can
+use it to make decisions on the client side.
+
+As an application developer, I want to be able to determine the
+Kubernetes distribution the application is being installed into at
```

Comment:
That's already available under `/version` endpoint or through `oc version` which reads just that.

Reply:
`oc version` shows the version, but not the distribution

```
$ oc version
Client Version: 4.11.0-0.nightly-2022-06-23-153912
Kustomize Version: v4.5.4
Server Version: 4.11.0-0.nightly-2022-06-23-153912
Kubernetes Version: v1.24.0+284d62a
```

Reply:
I added this detail to the body of the document.